File: arucopose.py
import cv2
import cv2.aruco as aruco 
import numpy as np 
import scipy 
from visualization.arucodetect import aruco_display
import json
import yaml
from pprint import pprint
from collections import defaultdict
from scipy.spatial.transform import Rotation as R 
import argparse
import glob
ARUCO_DICT = {
	"DICT_4X4_50": cv2.aruco.DICT_4X4_50,
	"DICT_4X4_100": cv2.aruco.DICT_4X4_100,
	"DICT_4X4_250": cv2.aruco.DICT_4X4_250,
	"DICT_4X4_1000": cv2.aruco.DICT_4X4_1000,
	"DICT_5X5_50": cv2.aruco.DICT_5X5_50,
	"DICT_5X5_100": cv2.aruco.DICT_5X5_100,
	"DICT_5X5_250": cv2.aruco.DICT_5X5_250,
	"DICT_5X5_1000": cv2.aruco.DICT_5X5_1000,
	"DICT_6X6_50": cv2.aruco.DICT_6X6_50,
	"DICT_6X6_100": cv2.aruco.DICT_6X6_100,
	"DICT_6X6_250": cv2.aruco.DICT_6X6_250,
	"DICT_6X6_1000": cv2.aruco.DICT_6X6_1000,
	"DICT_7X7_50": cv2.aruco.DICT_7X7_50,
	"DICT_7X7_100": cv2.aruco.DICT_7X7_100,
	"DICT_7X7_250": cv2.aruco.DICT_7X7_250,
	"DICT_7X7_1000": cv2.aruco.DICT_7X7_1000,
	"DICT_ARUCO_ORIGINAL": cv2.aruco.DICT_ARUCO_ORIGINAL,
}

marker_length = 0.052# unit : m ...
   

def camera_pose_board(corners, ids, K, dist, arucoDict,img):
    
    gap = 0.005
    markerLength = 0.052
    rvec = None
    tvec = None
    board = aruco.GridBoard((5,7),markerLength, gap, arucoDict)
    
    _, rvec, tvec= aruco.estimatePoseBoard(corners, ids, board, K, dist, rvec, tvec)
    
    Rmat, _ = cv2.Rodrigues(rvec.flatten())
    M = R.from_rotvec(np.pi*np.array([-1,0,0]))
    M = M.as_matrix()
	# estimatePoseBoard axis was flipped so modification will be required 
	# Make up rotation matrix

    cam_pos = -Rmat.T@tvec
    cam_ori = Rmat.T  
    cam_ori = cam_ori @ M
    
    
    return cam_pos , cam_ori
    
if __name__ =="__main__":
    # parser = argparse.ArgumentParser()
    # parser.add_argument("--type","-t",type=str,default="def")
    # args = parser.parse_args()
    
    # Init AruCo
    aruco_dict = ARUCO_DICT["DICT_6X6_250"] 

    with open("metadata/2023-05-03-23-11-22-camchain.yaml", 'r') as f:
        data = {}
        data = yaml.load(f,Loader= yaml.FullLoader)
        intrin = data['cam0']['intrinsics']
        
        dist = np.array((data['cam0']['distortion_coeffs']))
        K = np.array(((intrin[0], 0, intrin[2]), 
                    (0, intrin[1], intrin[3]), 
                    (0,         0,         1)), dtype = np.float32)                    
        
    # Camera pose Estimation witqh ArUco tags.
    ret = defaultdict(list)
    # init ArUco
    arucoDict = aruco.getPredefinedDictionary(aruco_dict)
    detectorParams = aruco.DetectorParameters()
    detector = aruco.ArucoDetector(arucoDict,detectorParams)

    images = sorted(glob.glob('output/images/test/*.png'))
    print("estimation proceedings..")
    for image in images:
        
        img = cv2.imread(image)
        
        h, w,_ = img.shape
        # Convert images to gray 
        gray =cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

        # Detect ArUco markers 
        corners ,ids, rejected = detector.detectMarkers(gray)   

        # visualize detected tags and origin
        origin_pixel = aruco_display(corners, ids, rejected, img)
        
        pos, ori = camera_pose_board(corners, ids, K, dist, arucoDict, img)
        
        cv2.imshow("res",img)    
        
        ret["poses"].append(pos.tolist())
        ret["orientations"].append(ori.tolist())

        key =cv2.waitKey(1) & 0xFF 
        if key == ord('q'):
            cv2.destroyAllWindows()


    print("- finished - ")
    with open('metadata/pos.json', 'w') as f:
        json.dump(ret,fp=f)
        

File: arucopose.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/aruco.hpp>
#include <opencv2/aruco/dictionary.hpp>
#include <opencv2/core.hpp>
#include <opencv2/core/types.hpp>
#include <opencv2/core/mat.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/calib3d.hpp>
#include <nlohmann/json.hpp>
#include <iostream>
#include <fstream>
#include <sstream>
using namespace std;
using namespace cv;
using json = nlohmann::json;

cv::Mat camera_pose_board(std::vector<std::vector<cv::Point2f>>& corners, std::vector<int>& ids, cv::Mat& K, cv::Mat& dist, cv::Ptr<cv::aruco::Dictionary> arucoDict, cv::Mat& img) {
    double gap = 0.005;
    double markerLength = 0.052;
    cv::Mat rvec, tvec;
    cv::Ptr<cv::aruco::GridBoard> board = cv::aruco::GridBoard::create(5, 7, markerLength, gap, arucoDict);
    
    cv::aruco::estimatePoseBoard(corners, ids, board, K, dist, rvec, tvec);
    
    cv::Mat Rmat;
    cv::Rodrigues(rvec, Rmat);
    cv::Mat M;
    cv::Rodrigues(cv::Vec3d(-CV_PI, 0, 0), M);
    
    cv::Mat cam_pos = -Rmat.t() * tvec;
    cv::Mat cam_ori = Rmat.t() * M;
    
    return cam_pos.clone(), cam_ori.clone();
}


int main(int argc, char** argv)
{
    CommandLineParser parser(argc, argv);
    parser.addArgument("--type", "-t", true);
    parser.setDefaultArgument("def");
    parser.parse();

    string type = parser.get<string>("type");

    // Init AruCo
    cv::Ptr<cv::aruco::Dictionary> arucoDict = cv::aruco::getPredefinedDictionary(cv::aruco::DICT_6X6_250);

    cv::FileStorage fs("/2023-05-03-23-11-22-camchain.yaml", cv::FileStorage::READ);
    cv::Mat K, dist;
    fs["cam0"]["intrinsics"] >> K;
    fs["cam0"]["distortion_coeffs"] >> dist;
    fs.release();

    // Camera pose estimation with ArUco tags
    vector<vector<cv::Point2f>> corners;
    vector<int> ids;
    vector<cv::Point2f> rejected;

    vector<string> images;
    glob("output/images/test/*.png", images);

    cout << "Estimation in progress..." << endl;

    for (const auto& image : images)
    {
        cv::Mat img = cv::imread(image);
        cv::Mat gray;
        cv::cvtColor(img, gray, cv::COLOR_BGR2GRAY);

        // Detect ArUco markers
        cv::aruco::detectMarkers(gray, arucoDict, corners, ids);

        // Visualize detected tags and origin
        cv::aruco::drawDetectedMarkers(img, corners, ids);
        cv::aruco::drawAxis(img, K, dist, cv::noArray(), cv::noArray(), 0.1f);

        cv::imshow("res", img);

        cv::waitKey(1);

        vector<cv::Vec3d> rvecs, tvecs;
        if (type == "def")
        {
            cv::aruco::estimatePoseSingleMarkers(corners, 0.1f, K, dist, rvecs, tvecs);
        }
        else if (type == "board")
        {
            cv::Mat board = cv::Mat::zeros(cv::Size(1, 1), CV_8UC1); // Placeholder for the board configuration
            pos , ori = camera_pose_board(corners, ids, K, dist, arucoDict)
        }

        json ret;
        for (size_t i = 0; i < rvecs.size(); ++i)
        {
            json pose;
            pose["translation"] = { tvecs[i][0], tvecs[i][1], tvecs[i][2] };
            pose["rotation"] = { rvecs[i][0], rvecs[i][1], rvecs[i][2] };
            ret["poses"].push_back(pose);
        }

        cv::destroyAllWindows();
    }

    cout << "- finished -" << endl;

    ofstream outputFile("pos.json");
    outputFile << ret.dump(4);
    outputFile.close();

    return 0;
}
